### PR TITLE
[react-native-navigation] Make navigator a class

### PR DESCRIPTION
**Pull request by danilobuerger**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/wix/react-native-navigation/blob/master/src/deprecated/indexDeprecated.ios.js#L14
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

---

Navigator is exported by rnn and is a class.
cc @egorshulga

### CODE REVIEW

**Comment by egorshulga** [comment]
What issues does this change solve?
A class without any implementation is effectively the same as an interface.

**Comment by danilobuerger** [comment]
@egorshulga I disagree. `Navigator` is explicitly exported and part of the public interface of that lib. Doing `new Navigator()` is legal so the types should reflect that. Also, it makes it very easy when testing a component that requires a navigator prop as you can just instantiate it:
```
// in some test.js
<SomeScreen navigator={new Navigator()} />
```

**Comment by egorshulga** [comment]
`Navigator` constructor accepts 3 parameters. Should type definitions reflect it as well?

**Comment by egorshulga** [comment]
Actually, your props definitions should not have `navigator` prop defined explicitly.
I'm just following the docs and examples, and the `Navigator` object is never instantiated and passed to a component.

**Comment by danilobuerger** [comment]
@egorshulga Of course it is passed into each pushed component ("Screen"). The library does that. See https://wix.github.io/react-native-navigation/#/usage?id=step-3-that39s-it

**Comment by egorshulga** [comment]
Yes, the library does this, and it has nothing to do with type definitions to be used by users.

**Comment by danilobuerger** [comment]
Alright a little example:
```
interface IProps {
  readonly navigator: Navigator;
}
export class SomeScreen extends React.Component<IProps> {
  // ...
}
```
If `SomeScreen` is pushed, rnn will set the `navigator`, everything works.
Now we want to test `SomeScreen`. So we have to pass in the `navigator` as we can't (not feasible anyway) involve rnn when doing a unit test. If `Navigator` was declared a `class` (which it clearly is) we can just do something like 
```
<SomeScreen navigator={new Navigator()} />
``` 
in our test.

**Comment by egorshulga** [comment]
You should not define 
`navigator` in `IProps`.
`NavigationComponentProps<IProps>` should be used instead.
This way you don't have to define `navigator` in your props but it is still accessible inside a component.

**Comment by egorshulga** [comment]
Besides, you cannot just pass a directly instantiated object in your test environment, cause in this case you have to override the `navigator` methods (so they produce some kind of noops, I presume).

**Comment by danilobuerger** [comment]
@egorshulga I disagree, and it's just a Union so the end result would be same. In either case, whether you use `Navigator` or `NavigationComponentProps` you still have to pass in the `navigator` in your test, else the test will fail if you use the navigator inside the component, for example:
```
TypeError: Cannot read property 'push' of undefined
```
Whether or not I have to override or mock some methods out in `Navigator` shouldn't be of concern for this typing.

Fact is: `Navigator` is publicly exposed and can be instantiated.
Fact is: The commit even states its for public use: "Export `Navigator` class for public use"

Why are you arguing against this? Different people will use the library in *legal* ways that you don't and that's good.

**Comment by egorshulga** [comment]
Ok, your reasons and explanation are convincing.
There is only one issue left. In your example `navigator` should be passed both in development and testing environments. 
But it should be optional outside the component and not-optional in the inside. How it can be achieved?

**Comment by danilobuerger** [comment]
I don't understand your question, sorry. Could you expand it a bit, please?

**Comment by egorshulga** [comment]
Both in tests and in regular code `this.props.navigator` is never `undefined`, that is why it's declaration is `navigator: Navigator`.

But when you instantiate a component there are two cases:
1) Regular code: `navigator` prop is optional, it should not be passed directly (cause it is created and maintained by the RNN lib).
2) Tests code: `navigator` prop is required, otherwise there would be `cannot read property ... of undefined` exceptions.

So my question is: how would you maintain both these cases?

PS: my opinion is the PR can be merged, but it raises a bunch of questions afterwards.

**Comment by egorshulga** [comment]
Possible solution could be to make `navigator` optional for both cases. But it clashes with the fact that `this.props.navigator` should always be defined from the inside of the component.
That is why I want to provide a unified way of injecting the `navigator` prop definition where it is needed, that is why `NavigationComponentProps` was introduced in the first place.

**Comment by danilobuerger** [comment]
I don't see the first case. When using RNN you normally don't instantiate the Screens yourself. For example, you would:
```
Navigation.registerComponent("foo", () => HomeScreen);
// ...
navigator.push({ screen: "foo" });
```
In my opinion, the navigator should always be required for screens. In regular code, you shouldn't instantiate it yourself and use the `navigator.xyz` methods. In testing, you should supply a mock or similar.

**Comment by egorshulga** [comment]
OK, I get it, you are right.

Then `NavigationComponentProps` should be changed like this:
```
export interface NavigationComponentProps {
    navigator: Navigator;
}
```
Could you add this change to your PR?

**Comment by danilobuerger** [comment]
@egorshulga added

**Review by egorshulga — APPROVED** [review]

File: types/react-native-navigation/react-native-navigation-tests.tsx [hunk 1]
```diff
@@ -2,7 +2,7 @@ import * as React from 'react';
 import { Text, View } from 'react-native';
 import { Navigation, NavigationComponentProps, NavigatorStyle, NavigatorButtons } from 'react-native-navigation';
 
-class Screen1 extends React.Component<NavigationComponentProps<{ height: number }>> {
+class Screen1 extends React.Component<NavigationComponentProps & { height: number }> {
```
Comment (DanielRosenwasser) [inline types/react-native-navigation/react-native-navigation-tests.tsx]: Would you mind just creating a `Screen1Props` that extends `NavigationComponentProps` and adds a `height: number`? Interfaces tend to display better, verify more, and are checked more efficiently in general.

**Comment by DanielRosenwasser** [comment]
Actually, instead of blocking this PR, feel free to send that update in the future.

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Thanks!

**Pull request merged**